HMIS links

eanders opened the pull request:
[//]: # 'remove this if PR is for a release-* branch'
# _Please squash merge this PR_

## Description

This:
1. Adds some checks to the links generated for linking off to HMIS to try and ensure someone will be able to access the item when they get there.
2. Standardizes the `to_warehouse` and `to_hmis` methods and makes it so you _can_ call them without needing a DB query

## Type of change
[//]: # 'remove options that are not relevant'

- [x] New feature (adds functionality)
- [x] Code clean-up / housekeeping

## Checklist before requesting review
- [x] I have performed a self-review of my code
- [x] I have run the code that is being changed under ideal conditions, and it doesn't fail
- [x] My code includes comments and/or descriptive variable names to help other engineers understand the intent (or not applicable)
- [x] My code follows the style guidelines of this project (rubocop)
- [x] I have updated the documentation (or not applicable)
- [x] If it's not obvious how to test this change, I have provided testing instructions in this PR or the related issue

eanders committed:
Don't show links to HMIS if you definitely don't have access to the object

gigxz commented on app/models/grda_warehouse/data_source.rb:
```diff
@@ -574,6 +574,33 @@ def hmis_url_for(entity)
 
+    # If we don't have the HMIS driver we probably aren't here, but we need to check for the next section
+    # If we don't have a user, just return the URl (backwards compatibility)
+    return url if user.blank? || ! HmisEnforcement.hmis_enabled?
```
This is ok but might be nicer to do
`return unless hmis? && HmisEnforcement.hmis_enabled?` on line 557 instead

eanders replied:
Will do!  Thanks.

eanders committed:
Move HMIS check to the top

gigxz commented on app/models/grda_warehouse/data_source.rb:
```diff
@@ -574,6 +574,33 @@ def hmis_url_for(entity)
+    # HMIS will handle access
+    return url if perms.blank?
+    # If we can see this in HMIS, don't bother linking to it
+    return nil unless hmis_user.permissions_for?(hmis_entity, perms, mode: :all)
```
@eanders this branch isn't working for me I wasn't seeing links anywhere even though I have access. I finally tracked down that this method fails silently if passed a `perms` array... it's supposed to be:

```suggestion
    return nil unless hmis_user.permissions_for?(hmis_entity, *perms, mode: :all)
```

With that fix, at least perms are attempting to resolve, but I get an error:

<img width="1023" alt="Screenshot 2024-03-28 at 4 23 23 PM" src="https://github.com/greenriver/hmis-warehouse/assets/5333982/74ef26ac-f2f7-4736-987c-6f9a7aa134bd">


Sorry that it wasn't so straightforward to figure out the access. I don't mind if this goes into release-110.

gigxz replied:
We have that `persisted?` check there as a regression check measure because we had some sort of bug around permissions on unpersisted records.  Maybe you can trick the model into knowing its persisted in `as_hmis` somehow..

gigxz commented on app/models/grda_warehouse/data_source.rb:
```diff
@@ -574,6 +574,33 @@ def hmis_url_for(entity)
+      'GrdaWarehouse::Hud::Project' => [:can_view_project],
+      # 'GrdaWarehouse::Hud::Organization' => [:can_view_organization],
+      'GrdaWarehouse::Hud::Client' => [:can_view_clients],
+      'Hmis::Hud::Enrollment' => [:can_view_projects, :can_view_enrollment_details],
```
```suggestion
      'GrdaWarehouse::Hud::Enrollment' => [:can_view_projects, :can_view_enrollment_details],
```
I dont think there is any place we are passing an HMIS enrollment to this? but I could be wrong

gigxz replied:
I'm specifically looking at https://hmis-warehouse.dev.test/source_data/59064?type=Enrollment which is a grdawarehouse model

eanders committed:
Reverting to always loading the HMIS/Warehouse record

eanders commented:
@gigxz sorry this was such a mess.  I _think_ it's sorted now.  I've reverted to always loading the associated record form the DB since that seems necessary for the permissions to work, and have, hopefully, cleaned things up a bit.

eanders committed:
Merge branch 'release-111' into ea/hmis-links-4844
